### DISCUSSION TITLE
Dynamic x-step between dates based on zoom levels

### DISCUSSION THREAD

**Discussion by OskarPersson** [patrykandpatrick/vico#553] [Ideas]
Based on the example shown [here](https://patrykandpatrick.com/vico/wiki/2.0.0-alpha.6/cartesian-charts/axes#value-formatting), how do I properly space out the values dynamically?

When adding more entries to the list of a longer period, it adds a entry to the x-axis for every day. Instead, the most optimal in my case would be to have the spacing be based on the current zoom level. So when zooming out the delta between to x-values increases.

Is this possible?

**Comment by patrickmichalik** [comment]
Hello! We plan on introducing this feature with Vico 2.0.0 Alpha 7, which we’re looking to release next week.

**Reply by OskarPersson** [reply]
Great! I see that this is now available. Do you have an example where I can see how that is used together with the Value formatting described [here](https://patrykandpatrick.com/vico/wiki/2.0.0-alpha.7/cartesian-charts/axes/#value-formatting)?

**Reply by OskarPersson** [reply]
@patrickmichalik 

I've tried updating Chart1 in the samples (based on 88286bec04c37d76ed77bc466e7fd1d6a3f55efc) to the following to enable dynamic spacing:

```patch
--- a/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart1.kt
+++ b/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart1.kt
@@ -72,12 +74,21 @@ private fun ComposeChart1(
                 rememberCartesianChart(
                     rememberLineCartesianLayer(axisValueOverrider = axisValueOverrider),
                     startAxis = rememberStartAxis(itemPlacer = startAxisItemPlacer),
-                    bottomAxis = rememberBottomAxis(guideline = null),
+                    bottomAxis = rememberBottomAxis(
+                        guideline = null,
+                        itemPlacer =
+                        remember {
+                            AxisItemPlacer.Horizontal.default(
+                                addExtremeLabelPadding = true,
+                            )
+                        }
+                    ),
                     persistentMarkers = remember(marker) { mapOf(PERSISTENT_MARKER_X to marker) },
                 ),
             modelProducer = modelProducer,
             modifier = modifier,
             marker = marker,
+            horizontalLayout = HorizontalLayout.fullWidth()
         )
     }
 }
``` 

This works great for the values defined in the sample, but if I increase the range, then the line is not visible at all anymore. 

```patch
--- a/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart1.kt
+++ b/sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/charts/Chart1.kt
@@ -104,6 +115,6 @@ private const val MAX_Y = 15f

 private val color1 = Color(0xffa485e0)
 private val chartColors = listOf(color1)
-private val x = (1..50).toList()
+private val x = (100..500).toList()
 private val axisValueOverrider = AxisValueOverrider.fixed<LineCartesianLayerModel>(maxY = MAX_Y)
 private val startAxisItemPlacer = AxisItemPlacer.Vertical.default(maxItemCount = { 6 })
```

**Reply by patrickmichalik** [reply]
Yup, the update is now available. We apologize for the slight delay. The problem you’re facing appears to be a bug—I’ve just resolved it, and a hotfix will be out within a few hours. I’ll let you know when it’s available, and I’ll prepare an example with dates along the _x_-axis.

**Reply by patrickmichalik** [reply]
@OskarPersson, the hotfix, [Vico 2.0.0 Alpha 8](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.8), is rolling out. Your modified version of the first sample chart should now work as expected. Beyond updating Vico, no changes are required on your end.

**Reply by OskarPersson** [reply]
Perfect, thank you!

**Reply by patrickmichalik** [reply]
@OskarPersson, here’s a quick example with dates:

```kt
val data = (1..31).associate { LocalDate.of(2024, 1, it) to Random.nextFloat() * 10 }
val xValuesToDates = data.keys.associateBy { it.toEpochDay().toFloat() }
val model = CartesianChartModel(LineCartesianLayerModel.build { series(x = xValuesToDates.keys, y = data.values) })
val dateTimeFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("d MMM")
val bottomAxisValueFormatter =
    AxisValueFormatter<AxisPosition.Horizontal.Bottom> { x, _, _ ->
        (xValuesToDates[x] ?: LocalDate.ofEpochDay(x.toLong())).format(dateTimeFormatter)
    }
val bottomAxisItemPlacer = AxisItemPlacer.Horizontal.default(addExtremeLabelPadding = true)
val horizontalLayout = HorizontalLayout.fullWidth()
```

```kt
CartesianChartHost(
    chart =
        rememberCartesianChart(
            rememberLineCartesianLayer(spacing = 64.dp),
            startAxis = rememberStartAxis(),
            bottomAxis =
                rememberBottomAxis(valueFormatter = bottomAxisValueFormatter, itemPlacer = bottomAxisItemPlacer),
        ),
    model = model,
    horizontalLayout = horizontalLayout,
)
```

https://github.com/patrykandpatrick/vico/assets/120058021/9c3d5ebe-6eb1-4107-9627-fa1b2f62de91

Because there can potentially be thousands of labels, not all labels are measured. This generally works fine, but for labels of unpredictable lengths, you might need to manually specify which labels should be measured. If you face any truncation issues, please let me know, and I’ll show you how to do this.
